### ISSUE TITLE
"kotlinx.serialization core version is too low" cumbersome message when previous version is present in transitive dependencies

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.serialization#993]
We have a lot of reports in slack with `Your current kotlinx.serialization core version is too low, while current Kotlin compiler plugin 1.4.0 requires at least 1.0-M1-SNAPSHOT. Please update your kotlinx.serialization runtime dependency` error message even though there is a direct dependency in the Gradle [[1]](https://kotlinlang.slack.com/archives/C7A1U5PTM/p1597840442067100).

The root of the problem -- the previous version is present in the transitive dependencies.

The proper solution would be to detect this transitivity and indicate it in the error message

**Comment by radzio** [comment]
It might be connected https://github.com/gradle/gradle/issues/14211

**Comment by netcyrax** [comment]
What if the transitive dependency is in a library that cannot be updated?

I have the same issue, not sure how to solve it.

**Comment by qwwdfsad** [comment]
>What if the transitive dependency is in a library that cannot be updated?

Then it's likely not compatible with Kotlin 1.4

**Comment by sandwwraith** [comment]
If the previous version is in the transitive dependency, it still won't work with 1.4, so this is not a bug.

Regarding error message, I'm not sure that we have infrastructure to track and print whole dependency tree.

**Comment by theromis** [comment]
Getting same issue after updating `kotlin-gradle-plugin` to `1.4.0` does workaround exists?

**Comment by sandwwraith** [comment]
@theromis Make sure that you do not have 0.20.0 version in your dependency tree via transitives

**Comment by theromis** [comment]
@sandwwraith already solved, in my case solution was just updating `kotlin` and `ktor` versions to set both to `1.4.0`
but after solving this issue `okhttp` stopped work on api `19`, think to fork ok http and comment out this line: https://github.com/square/okhttp/blob/master/okhttp/src/main/kotlin/okhttp3/internal/platform/AndroidPlatform.kt#L154
(because it works on 16, 17,18, 21...)

**Comment by xjcl** [comment]
For anyone stumbling across this, in my case I resolved this by replacing

    implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.20.0" 

with

    implementation "org.jetbrains.kotlinx:kotlinx-serialization-core:1.0.0-RC"

in `build.gradle`.

Can someone illuminate me on what the difference is? If this is just a renaming, can we trigger some sort of IDE warning about this?

**Comment by sandwwraith** [comment]
@xjcl 1.0.0 API is different with 0.20.0. Also, 0.20.0 is not compatible with Kotlin 1.4

**Comment by brett-eugenelabs** [comment]
> For anyone stumbling across this, in my case I resolved this by replacing
> 
> ```
> implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.20.0" 
> ```
> 
> with
> 
> ```
> implementation "org.jetbrains.kotlinx:kotlinx-serialization-core:1.0.0-RC"
> ```
> 
> in `build.gradle`.

This works for latest Android Studio Arctic Fox 2020.3.1 Beta 3 and Jetpack Compose.

**Comment by xjcl** [comment]
@sandwwraith I was curious about the renaming from `runtime` to `core`

**Comment by pdvrieze** [comment]
@xjcl the difference is mostly renaming, but it is also removing the json format (and other formats) from the core artefact (and putting them in their own artefact).

**Comment by DRSchlaubi** [comment]
I don't actually have an outdated version anywhere in my dependency tree: https://haste.schlaubi.me/ihehaconax.coffee
And still get it

**Comment by iromeo** [comment]
I get a similar error in 2021.3, but not in 2020.x.x: `Your current kotlinx.serialization core version is too low, while current Kotlin compiler plugin 1.6.0 requires at least 1.0-M1-SNAPSHOT. Please update your kotlinx.serialization runtime dependency.` error. Could be reproduced using recommended intellij plugin template (https://github.com/JetBrains/intellij-platform-plugin-template)

My minimal intellij plugin test case:
[kotlinx_issue.tar.gz](https://github.com/Kotlin/kotlinx.serialization/files/7643152/kotlinx_issue.tar.gz)

If you change `platformVersion` in `gradle.properties` to  `2021.2.3` - everyting is ok, but with `2021.3` compilation error.

**Comment by GlebIgnatevEmlid** [comment]
This issue needs more attention as it's pretty annoying.

@iromeo thanks for the workaround!

**Comment by segevfiner** [comment]
Derp. I have hit this in my plugin as well while trying to run it's tests with:
```
configurations.all {
    resolutionStrategy.sortArtifacts(ResolutionStrategy.SortOrder.DEPENDENCY_FIRST)
}
```

As it seems to otherwise not grab the dependencies in the tests in the same order as when doing `runIde`. No idea how to solve this to be able to run tests...

**Comment by cursive-ide** [comment]
Here's another vote for this bug being very annoying when building IntelliJ plugins. 

To make the workaround a little less intrusive, the annotation can be applied at the file level, by adding:
```kotlin
@file:Suppress("PROVIDED_RUNTIME_TOO_LOW")
```
to the top of your file. This makes it less onerous if you have tons of serialisable classes in a file.

**Comment by lulu2002** [comment]
hype, the problem still exists

**Comment by Flaxoos** [comment]
Yep, i'm also experiencing it, and also in the context of building intellij plugins

```
.../src/commonMain/kotlin/.../Something.kt:177:1 Your current kotlinx.serialization core version is 2.0.0, while current Kotlin compiler plugin unknown requires at least 1.0-M1-SNAPSHOT. Please update your kotlinx.serialization runtime dependency.
```
```
[versions]
java = "17"
kotlin = "2.0.0"
kotlinx-coroutines = "1.8.0"
kotlinx-coroutines-intellij = "1.8.0-intellij-11"
kotlinx-serialization = "1.7.1"
intellijPlatformPlugin = "2.2.1"
```

**Comment by cursive-ide** [comment]
Following on from this, if you're getting a similar error like:
```
Kotlin: Inline classes require runtime serialization library version at least 1.1-M1-SNAPSHOT, while your classpath has null.
```
Then the required magic incantation is:
```kotlin
@file:Suppress("PROVIDED_RUNTIME_TOO_LOW", "INLINE_CLASSES_NOT_SUPPORTED")
```

**Comment by talhakosen** [comment]
Problem:
You might see this error during build or in IDE:
"Your current kotlinx.serialization core version is too low, while current Kotlin compiler plugin 1.x.x requires at least 1.0-M1-SNAPSHOT."

Even if you added a direct dependency like:
    implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.6.2")
...you might still get this error. Why?

👉 Root Cause:
Older or incompatible versions of kotlinx.serialization (e.g. 1.3.x or 2.0.x) 
may be pulled in transitively via other dependencies — like Android Lint APIs or other tools.

Even a direct dependency may not override these effectively unless versions match *exactly* with the Kotlin plugin.

🛠️ Solution:
1. Apply the plugin explicitly:
   plugins {
       kotlin("plugin.serialization") version "2.1.20"
   }

2. Force correct kotlinx version explicitly:
   dependencies {
       implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.6.2")
   }

3. Use a resolution strategy (optional but recommended):
   configurations.all {
       resolutionStrategy {
           force("org.jetbrains.kotlinx:kotlinx-serialization-json:1.6.2")
           force("org.jetbrains.kotlinx:kotlinx-serialization-core:1.6.2")
       }
   }

4. Then run:
   ./gradlew clean build --refresh-dependencies --rerun-tasks

✅ After that, the @Serializable annotation should be processed correctly.

(If you *must* suppress temporarily: use @Suppress("PROVIDED_RUNTIME_TOO_LOW"))

**Comment by pdvrieze** [comment]
> 2. Force correct kotlinx version explicitly:
>    dependencies {
>    implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.6.2")
>    }
Please note that 1.6.2 is now a fairly old version and not designed for 2.1 support. You should go with a more recent version (where possible).

**Comment by xepozz** [comment]
Got the same error with plugin development.
Removed `compileOnly("org.jetbrains.kotlinx:kotlinx-serialization-json:1.8.1")` from `dependencies` and facing this.

Looks strange, because [Jakub Chrzanowski](https://platform.jetbrains.com/u/jakub.chrzanowski) said it was fixed: https://platform.jetbrains.com/t/compatibility-problems-compatibility-problems-detected/596/11
